### PR TITLE
Backend API

### PR DESCRIPTION
**Pull request by IcarusMJ12**

### CODE REVIEW

**Comment by megawidget** [comment]
Note PR is against my other branch for now

**Comment by megawidget** [comment]
Another thought: should I return `tf::Status` instead in the C++ API?

**Comment by aprocter** [comment]
> Another thought: should I return tf::Status instead in the C++ API?

Yup, I think that'd be a good idea.

File: python/ngraph/__init__.in.py [hunk 1]
```diff
@@ -71,3 +73,20 @@ def disable():
...
+
+def backends_len():
+  return ngraph.ngraph_backends_len()
+
+
+def list_backends():
+  len_backends = backends_len()
+  result = (ctypes.c_string * len_backends)(*(None * len_backends))
+  if not ngraph.ngraph_build_backends(result, len_backends):
+    raise Exception("Backends fluctuated while listing")
```
Comment (aprocter) [inline python/ngraph/__init__.in.py]: Do we expect this to happen?

File: test/python/test_api.py [hunk 2]
```diff
@@ -32,3 +34,17 @@ def test_disable(self, ngraph_device):
...
     assert ngraph_device.ngraph_is_enabled() == 1
+
+  def test_backends_len(self, ngraph_device):
+    assert ngraph_device.ngraph_backends_len()
+
+  def test_set_backend(self, ngraph_device):
+    assert ngraph_device.ngraph_set_backend(b"CPU")
+
+  def test_set_backend_invalid(self, ngraph_device):
+    assert ngraph_device.ngraph_set_backend(b"POTATO") == 0
```
Comment (aprocter) [inline test/python/test_api.py]: You haven't heard about Project POTATO?

Comment (megawidget) [reply test/python/test_api.py]: I have not :(

File: python/ngraph/__init__.in.py (hunk 1 above)
Comment (megawidget) [reply python/ngraph/__init__.in.py]: it seems very, very unlikely, but who knows

**Pull request merged**